### PR TITLE
Improve ExecutePureFunctions FuzzTests

### PR DESCRIPTION
**Pull request by StachuDotNet**
This PR addresses several issues and improvements in ExecutePureFunctions
- [x] The generator type present was unused, as the types did not match the property checked.
       Everything has since been moved to use RuntimeTypes
- [x] The ability to allow some fuzzer errors has been extracted into a separate `AllowedFuzzerErrors` module
- [x] Several generators have been extracted from the main generator, and code has been generally tidied up for ease of understanding/updates
- [x] JWT.signAndEncode are now included (marked as ImpurePreviewable)
- [x] The ProgramType-style property check previously used in ExecutePureFunctions has been copied, pasted, and tidied into StdLib.Tests.fs - I could not find a way to get these tests to pass with the new RT-centric property

While initially I had hopes of this PR addressing more of #3556, it's getting a bit hairy now, and would be best to merge with a separate PR addressing more of that work.

Additional improvements to be made:
- [ ] There is occassionally very notable sluggishness in both generators and in evaluations, likely created by:
  - `Gen.filter` usage where 'inverting' the generator' would be preferred (i.e. instead of "gen default" and then Gen.filter, write the generator from the ground up, without default being used)
  - "sized" generation combined with list generation sometimes yields very large nested lists/dicts
- [ ] String generation is currently somewhat broken. there are bad characters that are generated, often breaking tests
      To address this locally, I've been generating boring alphanumeric strings generally often, using some commented-out code alternatives in Generators.fs
- [ ] support more types and functions, per #3556

### CODE REVIEW

File: fsharp-backend/tests/FuzzTests/AllowedFuzzerErrors.fs [hunk 1]
```diff
@@ -0,0 +1,113 @@
+/// Error messages are not required to be directly the same between
+/// old and new implementations.
```
Comment (StachuDotNet) [inline fsharp-backend/tests/FuzzTests/AllowedFuzzerErrors.fs:2]: todo: add more context here

**Comment by StachuDotNet** [comment]
I'm currently working my way through resolving the RT v. PR generator/property difference. I'm proceeding assuming that RT everywhere is best - given that, adjusting the OCamlInterop to support an `executeExpr` similar to the `execute` fn. If this is the wrong way to go, please let me know.

**Comment by pbiggar** [comment]
> I'm currently working my way through resolving the RT v. PR generator/property difference. I'm proceeding assuming that RT everywhere is best - given that, adjusting the OCamlInterop to support an `executeExpr` similar to the `execute` fn. If this is the wrong way to go, please let me know.

RT everywhere seems right. However I seem to recall it's hard to convert OT to RT, so you might find it easier to do PT everywhere and convert to OT and RT as appropriate. I recall going back and forth between those choices (now that I think of it, I might have written a hacky RT->OT converter to deal with this).

**Comment by StachuDotNet** [comment]
FYI I managed to make everything `RT` without too much hassle. So far that seems to be working OK (still seeing issues elsewhere, but working though those)

**Comment by StachuDotNet** [comment]
Tidying this PR up for a review - additional work can be done separately, as this is already pretty large.

File: fsharp-backend/tests/allowedFuzzerErrors.jsonc [hunk 2]
```diff
@@ -1,34 +1,38 @@
...
     "Dict::toJSON_v1",
+
     // Sorting odities, these are fine
     "Dict::toList",
     "Dict::keys",
     "Dict::values",
-    // OCaml seems to trim wrong here,
-    "String::trim",
-    "String::trimEnd",
-    "String::trimStart",
```
Comment (StachuDotNet) [inline fsharp-backend/tests/allowedFuzzerErrors.jsonc:21]: I've found these to be working locally.

**Comment by StachuDotNet** [comment]
There were a few tests in StdLib.Tests.fs that used to reference the property checked in ExecutePureFunctions fuzztests.
I previously attempted to convert them to use the new RT-centric property, but got all various orders. To get this merge-able, with tests passing, I extracted/copied that property to the test file.

**Comment by StachuDotNet** [comment]
@pbiggar I'd like to cut the PR off here - there's more to do for #3556, but there's already a fair amount of code changes / improvements here. Would be best to address further enhancements in a separate PR. Ready for a round of review

**Review by pbiggar — COMMENTED** [review]
Seems good, though a lot of moving things around so a big tricky to tell. Happy to merge.

**Pull request merged**